src/cpp/rtps/builtin/data/WriterProxyData.cpp

```diff
@@ -101,400 +101,371 @@ WriterProxyData& WriterProxyData::operator=(const WriterProxyData& writerInfo)
     }
     if(m_qos.m_partition.sendAlways() ||  m_qos.m_partition.hasChanged)
     {
-        PartitionQosPolicy*p = new PartitionQosPolicy();
-        *p = m_qos.m_partition;
-        parameter_list.m_parameters.push_back((Parameter_t*)p);
+        if (!m_qos.m_partition.addToCDRMessage(msg)) return false;
     }
     if(m_qos.m_topicData.sendAlways() || m_qos.m_topicData.hasChanged)
     {
-        TopicDataQosPolicy*p = new TopicDataQosPolicy();
-        *p = m_qos.m_topicData;
-        parameter_list.m_parameters.push_back((Parameter_t*)p);
+        if (!m_qos.m_topicData.addToCDRMessage(msg)) return false;
     }
     if(m_qos.m_groupData.sendAlways() ||  m_qos.m_groupData.hasChanged)
     {
         GroupDataQosPolicy*p = new GroupDataQosPolicy();
         *p = m_qos.m_groupData;
-        parameter_list.m_parameters.push_back((Parameter_t*)p);
+        if (!m_qos.m_groupData.addToCDRMessage(msg)) return false;
     }
 
     if (m_topicDiscoveryKind != NO_CHECK)
     {
         if (m_type_id.m_type_identifier->_d() != 0)
         {
-            TypeIdV1 * p = new TypeIdV1();
-            *p = m_type_id;
-            parameter_list.m_parameters.push_back((Parameter_t*)p);
+            if (!m_type_id.addToCDRMessage(msg)) return false;
         }
 
         if (m_type.m_type_object->_d() != 0)
         {
-            TypeObjectV1 * p = new TypeObjectV1();
-            *p = m_type;
-            parameter_list.m_parameters.push_back((Parameter_t*)p);
+            if (!m_type.addToCDRMessage(msg)) return false;
         }
     }
 #if HAVE_SECURITY
     if ((this->security_attributes_ != 0UL) || (this->plugin_security_attributes_ != 0UL))
     {
-        ParameterEndpointSecurityInfo_t*p = new ParameterEndpointSecurityInfo_t();
-        p->security_attributes = security_attributes_;
-        p->plugin_security_attributes = plugin_security_attributes_;
-        parameter_list.m_parameters.push_back((Parameter_t*)p);
+        ParameterEndpointSecurityInfo_t p;
+        p.security_attributes = security_attributes_;
+        p.plugin_security_attributes = plugin_security_attributes_;
+        if (!p.addToCDRMessage(msg)) return false;
     }
 #endif
 
-    logInfo(RTPS_PROXY_DATA," with " << parameter_list.m_parameters.size()<< " parameters");
-    return parameter_list;
+    return CDRMessage::addParameterSentinel(msg);
 }
 
 bool WriterProxyData::readFromCDRMessage(CDRMessage_t* msg)
 {
-    ParameterList_t parameter_list;
-    if(ParameterList::readParameterListfromCDRMsg(msg, &parameter_list, NULL, true) > 0)
+    auto param_process = [this](const Parameter_t* param)
     {
-        for(std::vector<Parameter_t*>::iterator it = parameter_list.m_parameters.begin();
-                it!=parameter_list.m_parameters.end();++it)
+        switch (param->Pid)
         {
-            switch((*it)->Pid)
+            case PID_DURABILITY:
+            {
+                DurabilityQosPolicy * p = (DurabilityQosPolicy*)(param);
```

Comment:
Better use dynamic_cast instead of c-style casting.

Reply:
done